rocksdb/transactiondb.nim

```diff
@@ -107,22 +99,15 @@ proc isClosed*(db: TransactionDbRef): bool {.inline.} =
 
 proc beginTransaction*(
     db: TransactionDbRef,
-    readOpts = ReadOptionsRef(nil),
-    writeOpts = WriteOptionsRef(nil),
-    txDbOpts = TransactionDbOptionsRef(nil),
+    readOpts = defaultReadOptions(),
```

Comment:
if this is called without supplying readOpts/writeOpts, what happens? Who frees the options allocated by the default?

Reply:
ie in the colfamily pr, the idea was that if `nil` is given, the temporary options instance is "owned" by the callee (beginTransaction in this case) and therefor it also deallocates it

Reply:
They get freed when the db instance is closed. db.close() will free all the opts types regardless of whether passed in or not.

Reply:
Actually the above comment applies in the other locations. In this instance the opts are freed when the transaction is closed.

Reply:
what if the transaction/database/etc fails to open?

Reply:
In the case of beginTransaction, the database is already open and the call to rocksdb_transaction_begin doesn't have any error parameter so I guess its not expected to fail under normal conditions so no memory leak is possible in this case.

Reply:
In the case of opening the database then if there is an error we would have a memory leak but I guess in this case we would normally stop the program anyway. I can fix it by freeing the opt types in the bailOnErrors template.

Reply:
I think the core issue here is that when we do `xxx = defaultOptions` in the parameter list, the ownership is unclear which has these knock-on effects. semantically, it becomes strange that "sometimes" the function closes the passed-in options and sometimes it doesn't (if I pass in options, can I use them for another call? etc)

I'm not sure this is worth fixing, but it's certainly worth thinking about and the `nil` approach was a compromise (non-nil = caller owns) to get there.

Reply:
I actually found the solution using nil to be more unclear which is why I removed it. Using the default options solution makes it clear that the options will always get cleaned up when the database is closed. When using the nil method, the options require manual clean up if being passed in, if they don't get passed in then the options are cleaned up in the open function before the database is closed (this may cause other issues). The Java RocksDb library states ' Options instance *should* not be disposed before all DBs using this options instance have been closed', see here: https://javadoc.io/doc/org.rocksdb/rocksdbjni/latest/org/rocksdb/RocksDB.html#open-org.rocksdb.DBOptions-java.lang.String-java.util.List-java.util.List-

You should be able to use options for multiple calls as long as the database is open so for most of the functions that won't be a problem. In the case of beginTransaction these options get cleaned up when closing the transaction and so I guess this could be a problem if we want to reuse the opts across transactions. This function is not currently used in Nimbus by the way.

Reply:
I'm not sure if we can use nim destructors with the refc garbage collector but perhaps in a future change we can use destructors to help cleanup the opts types and then don't close any opt types in db.close().

Reply:
> I'm not sure if we can use nim destructors

no, these are ORC-only. we can indeed revisit the problem then - until then, perhaps it's worth documenting these quirks in the API as well, similar to java.

Reply:
Sure, I'll add that. 

I'm also working a change to support choosing if the types are freed when closing the database or not by adding a flag to each opt type. If you pass in an opt it won't get closed automatically but if none is passed in the default opts will get closed when the database is closed.